### PR TITLE
Feature #15664: Google OAuth plugin for emails

### PR DESCRIPTION
**Pull request by gabrieljenik**
WIP, Draft

### CODE REVIEW

**Comment by Shnoulle** [comment]
Thank you

**Comment by Shnoulle** [comment]
Looking at such plugin, and some other : i really think we need "Library plugin"/"vendor plugin" or something like that.

I have another plugin with paragonie, oauth etc … https://gitlab.com/SondagesPro/OrcidAuthenticate/-/tree/master/vendor

File: application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+    private function getHelpContent()
+    {
+        $this->subscribe('getPluginTwigPath');
+        $data = [
+            'redirectUri' => $this->getRedirectUri(),
+            'isHttp' => !(isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] === 'on'),
+        ];
+        return Yii::app()->twigRenderer->renderPartial('/Help.twig', $data);
+
+        // Translations here just so the translations bot can pick them up.
```
Comment (Shnoulle) [inline application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php:97]: :rofl:

File: application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+            [
+                'provider' => $provider,
+                'clientId' => $clientId,
+                'clientSecret' => $clientSecret,
+                'refreshToken' => $refreshToken,
+                'userName' => $emailAddress,
+            ]
+        ));
+
+        // Set "Reply To" because Gmail overrides the From/Sender with the logged user.
```
Comment (Shnoulle) [inline application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php:446]: :)

File: application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+        }
+
+        $emailAddress = $this->get("email");
+        if (empty($emailAddress)) {
+            $this->log("GoogleOAuthSMTP is enabled but there is no email address. Please generate a new token.", CLogger::LEVEL_WARNING);
+            return;
+        }
+
+        $limeMailer = $this->getEvent()->get('mailer');
+        $limeMailer->IsSMTP();
```
Comment (Shnoulle) [inline application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php:426]: Maybe adding 
```
if ($limeMailer->AuthType == 'XOAUTH2') {
    return;
}
```

(or more control)

File: application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+        $controller = $this->getEvent()->get('controller');
+        $action = $this->getEvent()->get('action');
+
+        if ($controller == 'admin' && $action == 'globalsettings') {
+            $this->registerGlobalSettingsScript();
+        }
+    }
+
+    private function registerGlobalSettingsScript()
+    {
```
Comment (Shnoulle) [inline application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php:469]: @olleharstedt about this hack to update Global Settings form, you don't work on something allowing update any form currently ?

Comment (olleharstedt) [reply application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php:469]: Yes, it's already available. https://manual.limesurvey.org/Plugins_-_advanced#Form_extension_.28New_in_6_.29

Might need to be extended, of course.

Comment (olleharstedt) [reply application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php:469]: But you can't use it to disable items, as they do here.

Comment (gabrieljenik) [reply application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php:469]: The idea is not to use a hack nor a plugin form event, but to enhance the core as to allow Oauth Email plugins.

**Comment by olleharstedt** [comment]
Hm current status?

**Comment by gabrieljenik** [comment]
This was a PoC.
Closed in favor of https://github.com/LimeSurvey/LimeSurvey/pull/3054

**Pull request closed without merging**